File: frontend/lib/models/scheme.dart
class SchemeDocument {
  final String docName;
  final String description;
  final bool isMandatory;
  String status; // 'Uploaded', 'Pending'

  SchemeDocument({
    required this.docName,
    this.description = '',
    this.isMandatory = true,
    this.status = 'Pending',
  });

  factory SchemeDocument.fromJson(Map<String, dynamic> json) {
    return SchemeDocument(
      docName: json['docName'] ?? '',
      description: json['description'] ?? '',
      isMandatory: json['isMandatory'] ?? true,
      status: json['status'] ?? 'Pending',
    );
  }

  Map<String, dynamic> toJson() => {
    'docName': docName,
    'description': description,
    'isMandatory': isMandatory,
    'status': status,
  };
}

class Scheme {
  final String id;
  final String schemeName;
  final String shortCode;
  final String category;
  final String tagline;
  final String description;
  final num maxGrantLoanAmount;
  final String loanAmountFormatted;
  final String interestRate;
  final double interestRateNumeric;
  final String repaymentPeriod;
  final int repaymentPeriodYears;
  final int minAge;
  final num maxIncome;
  final List<String> eligibleCategories;
  final List<String> eligibleBusinessTypes;
  final int minExperienceYears;
  final int subsidyPercentage;
  final String whoCanApply;
  final String purpose;
  final List<String> benefits;
  final List<SchemeDocument> requiredDocuments;
  final List<String> tags;
  final String applicationUrl;

  // Matching engine fields
  final int? matchPercentage;
  final String? matchBadge;
  final String? highlightTag;
  final List<String> eligibilityReasons;

  Scheme({
    required this.id,
    required this.schemeName,
    this.shortCode = '',
    this.category = 'Central Government',
    this.tagline = '',
    required this.description,
    required this.maxGrantLoanAmount,
    this.loanAmountFormatted = '',
    this.interestRate = '8% - 12%',
    this.interestRateNumeric = 9.5,
    this.repaymentPeriod = 'Up to 5 Years',
    this.repaymentPeriodYears = 5,
    this.minAge = 18,
    this.maxIncome = 0,
    this.eligibleCategories = const ['All'],
    this.eligibleBusinessTypes = const ['All'],
    this.minExperienceYears = 0,
    this.subsidyPercentage = 0,
    this.whoCanApply = 'Micro & Small Enterprises',
    this.purpose = 'Business Expansion, Working Capital, New Business',
    this.benefits = const [],
    this.requiredDocuments = const [],
    this.tags = const [],
    this.applicationUrl = '',
    this.matchPercentage,
    this.matchBadge,
    this.highlightTag,
    this.eligibilityReasons = const [],
  });

  factory Scheme.fromJson(Map<String, dynamic> json) {
    var rawScheme = json.containsKey('scheme') ? json['scheme'] : json;
    
    var docList = <SchemeDocument>[];
    if (rawScheme['requiredDocuments'] != null) {
      docList = (rawScheme['requiredDocuments'] as List)
          .map((d) => SchemeDocument.fromJson(d is Map<String, dynamic> ? d : {'docName': d.toString()}))
          .toList();
    }

    return Scheme(
      id: (rawScheme['_id'] ?? rawScheme['id'] ?? '').toString(),
      schemeName: rawScheme['schemeName'] ?? '',
      shortCode: rawScheme['shortCode'] ?? '',
      category: rawScheme['category'] ?? 'Central Government',
      tagline: rawScheme['tagline'] ?? '',
      description: rawScheme['description'] ?? '',
      maxGrantLoanAmount: rawScheme['maxGrantLoanAmount'] ?? 0,
      loanAmountFormatted: rawScheme['loanAmountFormatted'] ?? 'Up to ₹${rawScheme['maxGrantLoanAmount'] ?? 0}',
      interestRate: rawScheme['interestRate'] ?? '8% - 12%',
      interestRateNumeric: (rawScheme['interestRateNumeric'] ?? 9.5).toDouble(),
      repaymentPeriod: rawScheme['repaymentPeriod'] ?? 'Up to 5 Years',
      repaymentPeriodYears: rawScheme['repaymentPeriodYears'] ?? 5,
      minAge: rawScheme['minAge'] ?? 18,
      maxIncome: rawScheme['maxIncome'] ?? 0,
      eligibleCategories: List<String>.from(rawScheme['eligibleCategories'] ?? ['All']),
      eligibleBusinessTypes: List<String>.from(rawScheme['eligibleBusinessTypes'] ?? ['All']),
      minExperienceYears: rawScheme['minExperienceYears'] ?? 0,
      subsidyPercentage: rawScheme['subsidyPercentage'] ?? 0,
      whoCanApply: rawScheme['whoCanApply'] ?? 'Micro & Small Enterprises',
      purpose: rawScheme['purpose'] ?? 'Business Expansion, Working Capital',
      benefits: List<String>.from(rawScheme['benefits'] ?? []),
      requiredDocuments: docList,
      tags: List<String>.from(rawScheme['tags'] ?? []),
      applicationUrl: rawScheme['applicationUrl'] ?? '',
      matchPercentage: json['matchPercentage'],
      matchBadge: json['matchBadge'] ?? (json['matchPercentage'] != null ? '${json['matchPercentage']}% Match' : null),
      highlightTag: json['highlightTag'],
      eligibilityReasons: List<String>.from(json['eligibilityReasons'] ?? []),
    );
  }
}

File: backend/models/Scheme.js
const mongoose = require('mongoose');

const SchemeSchema = new mongoose.Schema({
  schemeName: {
    type: String,
    required: true,
    trim: true
  },
  shortCode: {
    type: String,
    trim: true
  },
  category: {
    type: String, // e.g., 'Central Government', 'State Government', 'MSME', 'Women Empowerment', 'Rural Development'
    default: 'Central Government'
  },
  targetSector: {
    type: String, // 'Agriculture', 'MSME / Small Business', 'Education / Youth', 'Women Entrepreneur', 'Artisans & Weavers', 'Street Vendors'
    default: 'MSME / Small Business'
  },
  state: {
    type: String, // 'Central', 'Telangana', 'Maharashtra', 'Andhra Pradesh', 'All'
    default: 'Central'
  },
  vernacularNames: {
    hi: { type: String, default: '' },
    te: { type: String, default: '' },
    kn: { type: String, default: '' },
    ta: { type: String, default: '' },
    mr: { type: String, default: '' },
    bn: { type: String, default: '' }
  },
  translations: {
    te: {
      tagline: { type: String, default: '' },
      description: { type: String, default: '' },
      benefits: [{ type: String }],
      whoCanApply: { type: String, default: '' },
      purpose: { type: String, default: '' },
      loanAmountFormatted: { type: String, default: '' },
      interestRate: { type: String, default: '' },
      repaymentPeriod: { type: String, default: '' },
      requiredDocuments: [{ docName: String, description: String }]
    },
    hi: {
      tagline: { type: String, default: '' },
      description: { type: String, default: '' },
      benefits: [{ type: String }],
      whoCanApply: { type: String, default: '' },
      purpose: { type: String, default: '' },
      loanAmountFormatted: { type: String, default: '' },
      interestRate: { type: String, default: '' },
      repaymentPeriod: { type: String, default: '' },
      requiredDocuments: [{ docName: String, description: String }]
    },
    kn: {
      tagline: { type: String, default: '' },
      description: { type: String, default: '' },
      benefits: [{ type: String }],
      whoCanApply: { type: String, default: '' },
      purpose: { type: String, default: '' },
      loanAmountFormatted: { type: String, default: '' },
      interestRate: { type: String, default: '' },
      repaymentPeriod: { type: String, default: '' },
      requiredDocuments: [{ docName: String, description: String }]
    },
    ta: {
      tagline: { type: String, default: '' },
      description: { type: String, default: '' },
      benefits: [{ type: String }],
      whoCanApply: { type: String, default: '' },
      purpose: { type: String, default: '' },
      loanAmountFormatted: { type: String, default: '' },
      interestRate: { type: String, default: '' },
      repaymentPeriod: { type: String, default: '' },
      requiredDocuments: [{ docName: String, description: String }]
    },
    mr: {
      tagline: { type: String, default: '' },
      description: { type: String, default: '' },
      benefits: [{ type: String }],
      whoCanApply: { type: String, default: '' },
      purpose: { type: String, default: '' },
      loanAmountFormatted: { type: String, default: '' },
      interestRate: { type: String, default: '' },
      repaymentPeriod: { type: String, default: '' },
      requiredDocuments: [{ docName: String, description: String }]
    },
    bn: {
      tagline: { type: String, default: '' },
      description: { type: String, default: '' },
      benefits: [{ type: String }],
      whoCanApply: { type: String, default: '' },
      purpose: { type: String, default: '' },
      loanAmountFormatted: { type: String, default: '' },
      interestRate: { type: String, default: '' },
      repaymentPeriod: { type: String, default: '' },
      requiredDocuments: [{ docName: String, description: String }]
    }
  },
  vernacularDetails: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  },
  embedding: {
    type: [Number], // 768-dimensional vector embedding for semantic search
    default: undefined
  },
  tagline: {
    type: String,
    default: ''
  },
  description: {
    type: String,
    required: true
  },
  maxGrantLoanAmount: {
    type: Number, // in INR
    required: true
  },
  loanAmountFormatted: {
    type: String, // e.g. "Up to ₹10,00,000"
    default: ''
  },
  interestRate: {
    type: String, // e.g., "8% - 12% (approx.)" or "Subsidized (0% - 7%)"
    default: '8% - 12%'
  },
  interestRateNumeric: {
    type: Number, // for financial calculation default, e.g. 9.5
    default: 9.5
  },
  repaymentPeriod: {
    type: String, // e.g., "Up to 5 Years"
    default: 'Up to 5 Years'
  },
  repaymentPeriodYears: {
    type: Number,
    default: 5
  },
  minAge: {
    type: Number,
    default: 18
  },
  maxIncome: {
    type: Number, // 0 or null means no income cap
    default: 0
  },
  eligibleCategories: [{
    type: String
    // 'General', 'OBC', 'SC', 'ST', 'Women Entrepreneur', 'Minority', 'All'
  }],
  eligibleBusinessTypes: [{
    type: String
    // 'Food Business', 'Handicrafts & Handlooms', 'Retail / Kirana Shop', etc. or 'All'
  }],
  minExperienceYears: {
    type: Number,
    default: 0
  },
  subsidyPercentage: {
    type: Number, // e.g. 15 to 35% subsidy
    default: 0
  },
  whoCanApply: {
    type: String,
    default: 'Micro & Small Enterprises'
  },
  purpose: {
    type: String,
    default: 'Business Expansion, Working Capital, New Business Setup'
  },
  benefits: [{
    type: String
  }],
  requiredDocuments: [{
    docName: { type: String, required: true },
    description: { type: String, default: '' },
    isMandatory: { type: Boolean, default: true }
  }],
  applicationUrl: {
    type: String,
    default: 'https://udyamregistration.gov.in'
  },
  tags: [{
    type: String // e.g. "Low Interest", "Collateral-Free", "High Subsidy", "Fast Process"
  }],
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true
});

module.exports = mongoose.models.Scheme || mongoose.model('Scheme', SchemeSchema);
